### ISSUE TITLE
Bug with repeated G28 within a macro?

### ISSUE THREAD

**Issue by mental405** [KevinOConnor/klipper#3272]
I have two macros configured to test the repeatability of two independent Z endstop switches. Attached to the frame are two digital dial indicators that I have hacked up to dump their data via a separate Arduino to a serial logger on a separate PC.

The intended behavior is: 

- Home Z Axis
- Move gantry up 100mm into the dial indicators
- Pull an output pin low to signal the arduino to capture the measurement and write it to the serial port
- Wait for the capture to complete
- Disable Z motors
- Repeat the test

The observed behavior is:
- Home Z Axis
- Move gantry up 100mm into the dial indicators
- Pull an output pin low to signal the arduino to capture the measurement and write it to the serial port
- Wait for the capture to complete
- Disable Z motors
- Repeat the test
- Z does not home a second time
- Z immediately begins moving up from current position to 100mm above current position (crashes)

For some reason when G28 Z0 is issued for a second time, it thinks the endstop pin is already triggered so it immediately sets the Z axis to 0 and begins moving. I traced the calls as best I could back to the ``_handle_endstop_state`` method in ``mcu.py``.

The initial homing move has the endstop state as 0 until the endstop is physically triggered
The subsequent homing move immediately reports the endstop as being triggered causing the undesirable crash.

I added ``QUERY_ENDSTOPS`` before the homing move and it behaves as intended. It is almost as if there is a flag that is not getting reset appropriately in the MCU and ``QUERY_ENDSTOPS`` is clearing it. 

I really just want to know if anyone else can observe this behavior. I have not attached a klippy.log because I didn't see anything relevant with the endstop states within it, only after redirecting the output [here](https://github.com/KevinOConnor/klipper/blob/db05938a5fd04206ec297c816f5f7fbd0a86e421/klippy/mcu.py#L77) to the log file was I able to see that the endstop was reporting triggered at the onset of the move.

The macros have been simplified a bit but the behavior can be observed here.
```
[gcode_macro TEST_Z_REPEATABILITY]
variable_test_sample:             5
variable_test_total:              0
gcode:
    #dirty hack fix 
    QUERY_ENDSTOPS

    #home z
    G28 Z

    #move z to dial indicators
    G1  Z100

    #tell sensor to capture data by pulling the
    #pin low for 100ms and wait for capture
    G4  P100
    G4  P1000

    #reset
    {% if S < T %}
        SET_GCODE_VARIABLE MACRO=TEST_Z_REPEATABILITY VARIABLE=test_sample VALUE={S|int}
        UPDATE_DELAYED_GCODE ID=repeat_test_z_repeatability_repeatedly DURATION=2
    {% endif %}

[delayed_gcode repeat_test_z_repeatability_repeatedly]
initial_duration: 0
gcode:
    M18
    TEST_Z_REPEATABILITY  
```

**Comment by klipper-gitissuebot** [comment]
It appears the Klipper log file was not provided.
I'm closing this now, sorry. Feel free to attach the log to this ticket sometime in the future, or create a new issue once you can provide the log file.
